internal/searchindex/searchindex.go

```diff
@@ -0,0 +1,440 @@
+package searchindex
+
+import (
+	"bytes"
+	"encoding/gob"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/cmp"
+
+	"go.mongodb.org/atlas-sdk/v20231115008/admin"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/compat"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/common"
+)
+
+type SearchIndex struct {
+	akov2.SearchIndex
+	akov2.AtlasSearchIndexConfigSpec
+	ID     *string
+	Status *string
+}
+
+func (s *SearchIndex) Key() string {
+	var buf bytes.Buffer
+	encoder := gob.NewEncoder(&buf)
+	err := encoder.Encode(*s)
+	fmt.Println("ERR:", err)
+	return buf.String()
+}
+
+func (s *SearchIndex) GetID() string {
+	return pointer.GetOrDefault(s.ID, "")
+}
+
+func (s *SearchIndex) GetStatus() string {
+	return pointer.GetOrDefault(s.Status, "")
+}
+
+func NewSearchIndexFromAKO(index *akov2.SearchIndex, config *akov2.AtlasSearchIndexConfigSpec) *SearchIndex {
+	if index == nil || config == nil {
```

Comment:
hm, this is a hard invariant. Please document why this is necessary and why a partially constructed object has no meaning, or, allow partial construction.